### PR TITLE
HTML Structure

### PR DESCRIPTION
**Pull request by cpfeifer**
### Pull Request Goal
Create default template positions, baseline html structure in index.php and baseline module position rendering.

With this in place, we can begin working on backend architecture and parameters, template functionality and template styling.

### Summary of Changes
New index.php
- basic html5 structure and position rendering
- no CSS has been applied, only HTML structures and jdoc includes for new wireframe concept

Added new positions in templateDetails.xml

Added HTML Folder
- Added html/modules.php
- Included html/layouts and chrome from Protostar for reference

### Testing Instructions
Each position should render only if a module is assigned
Currently there is no styling or column control, all positions and modules will stack on top of each other

### Documentation Changes Required
None yet

### CODE REVIEW

**Comment by designbengel** [comment]
i would remove the hardcoded parts and use module chrome for output. you can use the moduletag in chrome to define if its aside section div or something else. Can do a PR this week on this if you want...

**Comment by cpfeifer** [comment]
Thanks @designbengel 

We can definitely slim it down, this is just a starting point. Containers and rows can come from elsewhere, and the more we can consolidate the better. I'll go after the module chrome and rendering next. Maybe we can create a jlayout as well, or something along those lines. A class based approach will offer more options than module functions, but I don't want to overcomplicate it if we don't need to.

I'd love to see any ideas you have :)

**Comment by brianteeman** [comment]
Using php to write comments inside html just seems wrong to me
> <?php // end sidebar-right ?>

**Comment by cpfeifer** [comment]
@brianteeman - one advantage is they never show in the source code output. I'm not overly attached to that particular idea, and I'm sure there are many other standard nomenclature inconsistencies in here as well. 

The goal of this PR is to decide on the template structure, and it is admittedly a very rough outline. I can address the finer points once we decide on the direction. Just getting the ball rolling.

**Comment by cpfeifer** [comment]
As reference, this the basic wireframe concept I'm working with. It's a little more than a basic layout, but we could a lot with this and it gives users some very flexible customization options.

tauristar-wireframe

File: templates/tauristar/html/layouts/joomla/form/field/contenthistory.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<?php
```
Comment (laoneo) [inline templates/tauristar/html/layouts/joomla/form/field/contenthistory.php]: Why the form field overrides?

**Comment by laoneo** [comment]
I agree with @brianteeman, I would remove the PHP comment tags. The structure is so simple, no need for them.

**Comment by cpfeifer** [comment]
@laoneo Comments removed

The layouts can go. I just copied the html folder from protostar because there was not an html folder or modules.php to work with. I'm not sure how I can delete files, but it's not there for any particular reason other than reference. 

This is a pretty simple PR, the current index is un-usable and we're missing the HTML folder & module chrome. Let's just get it in and start working.

**Comment by designbengel** [comment]
I would prefer to follow our first plan to first define the parameters the user can set and then create the index from these settings. You can´t build the index without planning the parameters before IMHO. Also the right chromes should be used.

**Comment by cpfeifer** [comment]
We still need figure out the chroming / section layout piece of this. We don't want to use the chrome to generate sections or page structure, it's purpose is to generate a module wrapper. The chrome needs to load inside of an existing structure, this html provides that, but I think a layout would be better for this purpose - maybe "jlayout.template.position" or something. 

The best way to do that depends on the page structure we're using, so we need to decide on a page structure, hence this PR.

We can use this layout to move forward on everything. I can use this layout to build an interface, you can use this layout to start working on designs, the developers can use it to building the layout / chroming functions and additional programming and so on. 

This is not meant to be a finished product, we still have a lot of work to do. We need a foundation we can all work with, and we do not have that currently. We have to start somewhere and this covers all the basics.

**Comment by designbengel** [comment]
>  We don't want to use the chrome to generate sections or page structure

What do you mean with that? That´s exactly what a chrome is for?

**Comment by cpfeifer** [comment]
That is not correct, module chrome does not generate the page structure. It generates an html structure around an individual module, regardless of where it is on the page. 

For example, we don't want to add a "row" or <section> to every module position, but we do want to use it to add a class name, or bootstrap parameters, which set how many columns are in the section. The sections are not part of the chrome, the chrome goes inside the sections. It doesn't work properly otherwise.

**Comment by cpfeifer** [comment]
From a technical standpoint, chrome is a function not an object. The only parameters that get passed to it are module parameters passed by reference. We can't pass the template parameters from the index file to it easily, because those are part of the page object. We can use "this->" in the index file, but in not modules.php

Chroming is one of the few things left in Joomla that is not an object, it could certainly be developed into one, but that's way outside the scope of this PR and it's not necessary for our current purposes. We do need to address the structural piece, but chrome is not the answer in this case.

**Comment by designbengel** [comment]
sorry, i do not agree on this...
the module tag parameter and on aren't there for fun purpose... ;-)

**Comment by cpfeifer** [comment]
It's more of a fact than an opinion. 

Try putting those two lines of html in the chrome function, then create a section with 3 columns using only module parameters. Even if you can get it to display properly, it would take a lot of unnecessary CSS and it would introduce an entirely new set of problems. I can save you some time by telling you it won't work but feel free to try if you don't believe me.

We need something more than the existing module chrome system to do this properly. The chrome is one piece of the puzzle. There is another structural piece we need outside of the chrome to create the page structure. 

Maybe we do need to address the chrome function in some way and improve it, but that shouldn't be holding up the addition of working code and missing files. We don't have a modules.php in the current build to work with. It has to exist before we can work with it, so let's add it and move forward.

**Comment by cpfeifer** [comment]
From the Bootstrap documentation: http://getbootstrap.com/css/#grid

The first 4 bullet points:
- Rows must be placed within a .container (fixed-width) or .container-fluid (full-width) for proper alignment and padding.
- Use rows to create horizontal groups of columns.
- Content should be placed within columns, and only columns may be immediate children of rows.
- Predefined grid classes like .row and .col-xs-4 are available for quickly making grid layouts. 

The chrome is bullet point #4, and it's only possible if the other 3 are already in place. There has to be a surrounding container and row for the BS column CSS classes to even work. It is not possible to have a single module chrome create a container for itself or other modules, or for itself and other modules. 

The containers in this PR are surrounded by conditionals, they only load if there is a module published in that position. There are more efficient ways to deal with repeating code, but it's a  minimal amount code that is required, and there is nothing wrong with doing it like this. Any performance difference is going to minimal and building any solution for creating a container div will take longer than typing it 10 times. It's a minor detail we can address later in my opinion.

**Comment by laoneo** [comment]
We used some chrome stuff in pr #101. Perhaps it helps a bit.

**Comment by mbabker** [comment]
What are you building, the Joomla Template Framework™ or a simple starting point for CMS users?  If you're going for the former, you're going to end up with an overly bloated and overly complex template.  If you're going for the latter, it's by definition going to be a lot simpler and probably harder to do all sorts of fancy stuff with, but do we honestly need an overly complex template in the core?  Remember we shipped an Atomic Framework template with 1.6 thru 2.5 and look how popular that was.

**Comment by brianteeman** [comment]
I've been wondering the same

**Comment by cpfeifer** [comment]
The idea is to have a template that's usable and customizable, and show people what is possible with the joomla template system and core functionality.

It's not as complicated as it looks. We can create one piece of code for each section that can be re-used for all of the section on the front and back end. This also creates new ways for template developers to extend using core functionality - which is currently a rare phenomenon. 

If we can build the right architecture and provide options that rival commercial options, we can keep more people on the core and start to minimize third party issues by giving developers something they can work with. This is all possible right now and we're not taking advantage of it, and it's hurting us in my opinion.

**Comment by mbabker** [comment]
Who's your target user?  For the inexperienced user who probably couldn't make a difference between HTML, CSS, and JavaScript, this might work for them.  Anyone with a skill set beyond that, I just see code bloat.

I have to suffer through a lot of sites built with Gantry and for the simplest of tasks I spend far too much time reverse engineering their code bloat and over-complexity when making changes.  In the items I see on this repository, I honestly see a lot of over-complexity being introduced with minimal gain.

**Comment by cpfeifer** [comment]
My target is everyone. We can make an interface that's easy enough for end users to personalize in a useful way on top of a flexible code base that integrators and developers can customize. Currently, we have neither of those. The first step in Joomla after install for many users is installing a third party template. It's the first thing people see on the front end after install, and if it's something they can work with, it only helps us.
 

The last thing I want is to make it too complicated for anyone. We can leverage what Joomla already does to create a template system that appeals to our user base. We don't have to re-invent the wheel to do this. 

I'll build out what I'm thinking. It's hard to explain, but I know it will work and it will be easy to maintain. I'd rather spend time working on it then debating if it's it's necessary or not, because I already believe it is it necessary and I know it will help the platform.

**Comment by mbabker** [comment]
> My target is everyone

And that's where you're doomed to fail.  That's why Joomla has so many issues; if you try to please everyone you're going to please no one.  You NEED to set a clear target and a clear purpose beyond "build it so everyone can use it for all the things".

> The first step in Joomla after install for many users is installing a third party template

So?  Do we really need to structure the default front-end template in such a way that everyone wants to use it?  And you're right, our procedure at the office on new sites is to set up a template; usually one we have built from scratch using our own media and markup structures.  Bootstrap does not cut it for the sites that we build designs for.  No matter what you do to the default template that is not going to fix our use cases.

> The last thing I want is to make it too complicated for anyone. We can leverage what Joomla already does to create a template system that appeals to our user base. We don't have to re-invent the wheel to do this.

Except you're on that path.  Parameters to change the CSS framework?  Dynamic module positions?  These are features I find in the overly bloated template frameworks that I can't get removed from sites quick enough, not the more simple structure of the default templates shipped with the core distro.  I don't think the default template shipped with Joomla needs to address every user's potential needs, that's the direction I see you going here.

Joomla already has a mostly logical template system.  The worst things about it are knowing what you have to override and where.  It's not easy for users to know what layouts in com_content to override if they want to customize the category views, or that unless you place things just right overridding a JLayout is a global thing even if you only want it for one view or extension.  Building a complex template isn't going to change that.

> We can make an interface that's easy enough for end users to personalize in a useful way on top of a flexible code base that integrators and developers can customize.

I don't want a template framework in core that makes it more mandatory to use the interface to customize it.  These third party template frameworks make your life a living hell if you want to customize certain aspects of the template; these are things I (and many others who grasp PHP and the structure of Joomla's system) can easily customize in templates by way of overrides or building our own tools.

**Comment by AlisonAMG** [comment]
Bloat - no thanks. Will this be able to be removed? 

I always leave Protostar installed to test against if I'm having an issue. This sounds like it's going to weigh things down. Please don't.

**Comment by laoneo** [comment]
We appreciate your concerns. To really help Joomla, would be nice to create a PR here how the files index.php and templateDetails.xml should look like in your opinion. Without seing some examples it is difficult to move forward. When I look on the code of protostar, then it is IMO not less complex than tauristar.

**Comment by cpfeifer** [comment]
I'm just trying to move things forward. I'll fill in the blanks and dial it in more.

There isn't aspect of this I haven't considered, not all of is covered in the PR, but I know what needs to be done. This is more of an approach than a finished product. It's not nearly as complicated as might appear and it can be made simpler in many ways.

I know what I'm proposing is possible, and I know it can be more useful and easier to work with for end-users and for developers than what we have currently.

**Pull request closed without merging**